File: src/NerdBank.GitVersioning/CloudBuildServices/VisualStudioTeamServices.cs
// Copyright (c) .NET Foundation and Contributors. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

namespace Nerdbank.GitVersioning.CloudBuildServices;

/// <summary>
/// Cloud build handling for Azure DevOps.
/// </summary>
/// <remarks>
/// The VSTS-specific properties referenced here are <see href="https://msdn.microsoft.com/en-us/Library/vs/alm/Build/scripts/variables">documented here</see>.
/// </remarks>
internal class VisualStudioTeamServices : ICloudBuild
{
    /// <inheritdoc/>
    public bool IsPullRequest => BuildingRef?.StartsWith("refs/pull/") ?? false;

    /// <inheritdoc/>
    public string BuildingTag => CloudBuild.IfStartsWith(BuildingRef, "refs/tags/");

    /// <inheritdoc/>
    public string BuildingBranch => CloudBuild.IfStartsWith(BuildingRef, "refs/heads/");

    /// <inheritdoc/>
    public string GitCommitId => null;

    /// <inheritdoc/>
    public bool IsApplicable => !string.IsNullOrEmpty(Environment.GetEnvironmentVariable("SYSTEM_TEAMPROJECTID"));

    private static string BuildingRef => Environment.GetEnvironmentVariable("BUILD_SOURCEBRANCH");

    /// <inheritdoc/>
    public IReadOnlyDictionary<string, string> SetCloudBuildNumber(string buildNumber, TextWriter stdout, TextWriter stderr)
    {
        (stdout ?? Console.Out).WriteLine($"##vso[build.updatebuildnumber]{buildNumber}");
        return GetDictionaryFor("Build.BuildNumber", buildNumber);
    }

    /// <inheritdoc/>
    public IReadOnlyDictionary<string, string> SetCloudBuildVariable(string name, string value, TextWriter stdout, TextWriter stderr)
    {
        Utilities.FileOperationWithRetry(() =>
            (stdout ?? Console.Out).WriteLine($"##vso[task.setvariable variable={name};]{value}"));
        return GetDictionaryFor(name, value);
    }

    private static IReadOnlyDictionary<string, string> GetDictionaryFor(string variableName, string value)
    {
        return new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase)
        {
            { GetEnvironmentVariableNameForVariable(variableName), value },
        };
    }

    private static string GetEnvironmentVariableNameForVariable(string name) => name.ToUpperInvariant().Replace('.', '_');
}

File: src/nerdbank-gitversioning-rs/src/cloud_build_services/visual_studio_team_services.rs
// Copyright (c) .NET Foundation and Contributors. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

//! [Azure DevOps](https://learn.microsoft.com/azure/devops/pipelines/build/variables) support.

use std::io::{self, Write};

use crate::cloud::{
    CloudBuild, EnvironmentVariables, env, if_starts_with, is_nonempty, variable_for_environment,
};

/// Azure DevOps (formerly Visual Studio Team Services) integration.
pub struct VisualStudioTeamServices;

impl CloudBuild for VisualStudioTeamServices {
    fn name(&self) -> &'static str {
        "VisualStudioTeamServices"
    }

    fn is_applicable(&self) -> bool {
        is_nonempty(env("SYSTEM_TEAMPROJECTID").as_deref())
    }

    fn is_pull_request(&self) -> bool {
        env("BUILD_SOURCEBRANCH").is_some_and(|value| value.starts_with("refs/pull/"))
    }

    fn building_branch(&self) -> Option<String> {
        if_starts_with(env("BUILD_SOURCEBRANCH"), "refs/heads/")
    }

    fn building_tag(&self) -> Option<String> {
        if_starts_with(env("BUILD_SOURCEBRANCH"), "refs/tags/")
    }

    fn git_commit_id(&self) -> Option<String> {
        None
    }

    fn set_cloud_build_number(
        &self,
        build_number: &str,
        stdout: &mut dyn Write,
        _stderr: &mut dyn Write,
    ) -> io::Result<EnvironmentVariables> {
        writeln!(
            stdout,
            "##vso[build.updatebuildnumber]{}",
            escape_data(build_number)
        )?;
        Ok(variable_for_environment("Build.BuildNumber", build_number))
    }

    fn set_cloud_build_variable(
        &self,
        name: &str,
        value: &str,
        stdout: &mut dyn Write,
        _stderr: &mut dyn Write,
    ) -> io::Result<EnvironmentVariables> {
        let escaped_name = escape_property(name);
        let escaped_value = escape_data(value);
        writeln!(
            stdout,
            "##vso[task.setvariable variable={escaped_name};]{escaped_value}"
        )?;
        writeln!(
            stdout,
            "##vso[task.setvariable variable={escaped_name};isOutput=true;]{escaped_value}"
        )?;
        Ok(variable_for_environment(name, value))
    }
}

fn escape_data(value: &str) -> String {
    value
        .replace('%', "%AZP25")
        .replace('\r', "%0D")
        .replace('\n', "%0A")
}

fn escape_property(value: &str) -> String {
    escape_data(value).replace(']', "%5D").replace(';', "%3B")
}
